### DISCUSSION TITLE
Minimum pulse width in GRBL

### DISCUSSION THREAD

**Discussion by rcecinati** [winder/Universal-G-Code-Sender#2304] [Help]
Hello everybody,
I am starting with my new DIY CNC router, and I am facing  a weird problem. My setup is as follows:
Arduino UNO clone
CNC Shield 3
3 x A4988 stepper drivers 
3 x 42SHDB 4036Z-24B stepper motors
Power supply: 12v DC, 10A, running both the Arduino and the shield.
My sw setup is Arduno IDE on a Win 11 laptop.
I'd like to run the toy with UGS , it  looks like the best choice to me. 

The issue: running UGS,  motors  get locked after connecting, no setup wizard is possible. 
Warning:  after connection, a red ALARM is shown on the bottom of UGS, but with  no further hint

After trimming the three VRef on all of the drivers (0.4V) , I first managed to run the motors, following the valuable guide:
Demos that exercise the CNC Shield v3.xx directly (no GRBL layer or such) (https://github.com/MartyMacGyver/CNC_Shield_Demos, thank you Marty). I succeeded to run the test, but  only changing the minimum pulse duration from 350 us to 400 us or above

So I loaded GRBL (1.1f) on Arduino IDE, compiled and uploaded. Everything smooth.
Turn Arduino IDE off (to release the serial USB channel), run UGS Platform (2.0.14), verify channel port (COM3), speed (115200), firmware (GRBL). After connecting, red Alarm turns on (undecipherable), and the steppers get a faint movement. After that , nothing.
I noticed from instructions :
Grbl v1.1 Configuration  (https://github.com/gnea/grbl/wiki/Grbl-v1.1-Configuration#quick-guide-to-setting-up-your-machine-for-the-first-time), that the parameter $0  (minimum pulse width, us) is defaulted to 10 us, which is well below the value of 400 us I empirically found in the GRBL-less setup. But if I try to raise that value, I find that it's trimmed to 256,  and you can't  go beyond that.
I guess that GRBL casts that value to 1 byte.
In the A4988 datasheet, I read that the minimum pulse width is Th=1 is, Tl=1us , so the $0 setting appears to be right. So, why the hell my A4988 appear to need so a long pulse time?
And, how do I know what the UGS Alarm signal is trying to tell me?  (really don't  know if the two things are tied).

Thank you for helping, it's been teasing me for days...
Riccardo

**Comment by breiler** [comment]
There is a lot to go through here and it is too bad that you couldn't get the Setup Wizard running. It could help to resolve some issues. You could try the latest version of UGS 2.0.20 from here and see if that makes a difference: https://github.com/winder/Universal-G-Code-Sender#downloads

The alarm when you are starting the machine is basically GRBL telling you it doesn't know its position and require you to home the machine. You don't need to set up homing just yet, so you could just unlock the machine (see [Unlock](https://github.com/winder/Universal-G-Code-Sender/wiki/Usage#actions)). Now issuing commands or jogging should be possible. If you are getting tired to unlocking your machine everytime you connect you can disable homing `$23`. 

I find that the default values for `$0` and `$1` is fine for getting started with. You can always tweak these later.

If your machine starts moving after unlocking you are ready for the next configuration steps:
1.  Ensure the direction of each axis `$3` 
2.  Calibrate the steps per millimeters `$100-$102`
3.  Find out your max rate `$110-$112`

After all of these you can go on and figure out limit switches, homing, soft limits, probing and so on.

**Reply by rcecinati** [reply]
dear Breiler,
thank you for your response. 
I've tried the 2.0.20 release,  and nothing seems to change. What do you think of the Minimum Pulse Width setting? it appears puzzling to me.  Is it possible to force a 400+  use pulse in GRBL, forcing the 255 usec casting limit?
Is it necessary to install a software driver named  CH341SER.exe such as 
https://docs.sainsmart.com/article/xevbm8qufa-how-to-install-universal-gcode-sender-ugs-for-windows-operating-system
I've installed, but with no luck
Of course I've not mentioned the dozens of guides and sites giving clues and directions, they're driving me crazy.
I'm ordering a new kit of hardware stuff (Arduino, shield, A488s) just in case of some hw failure:
https://www.banggood.com/it/TWO-TREES-CNC-Shield-+-UNO-R3-Board-+4x-A4988-Stepper-Motor-Driver-+4x-4401-Stepper-Motor-Kit-for-3D-Printer-p-1510212.html?rmmds=myorder

Thank you for your hints
Riccardo

**Reply by breiler** [reply]
> Is it necessary to install a software driver named CH341SER.exe such as

It is necessary if you have an Ardunio with that serial chip. Impossible to answer without knowing exactly which Arduino you bought. But as you were able to connect and get the alarm message your serial connection seems to function and you can connect and communicate with the controller.

> What do you think of the Minimum Pulse Width setting?

Forget about the pulse width for a moment. If you still have an Alarm from the controller you won't be able to make the machine move. Did you try and unlock it?

**Reply by rcecinati** [reply]
dear Joacim
you're so fast.  No change after unlock, apart from ALARM turning into "Idle".  if then I issue, for instance, a "home machine", status turns into "HOME" and motors give a bump, after which everything goes asleep until next Disconnect/Connect
I've ordered new hardware

Thank you

**Reply by breiler** [reply]
As I wrote before, homing comes later. Don't bother with it until you have basic movement.
Plain jogging will suffice.

Make note of machine coordinates in the DRO when jogging. They should change. If you have soft limits activated ($20) it will prevent movement outside the machine boundaries.

Looking through what you wrote, you are now running GRBL 1.1 on a "CNC Shield 3". I have never tried those, but they usually require GRBL 0.9. Look through its specification.

Post your settings from the $$ command.

**Reply by rcecinati** [reply]
Joacim,
I am having difficulties in finding a valid 0.9 .zip release for GRBL. I didn't know of mismatches between GRBL and CNC Shields. I'll let you know.
btw, my current $$ setting is as follows:
>>> $$
$0 = 10    (Step pulse time, microseconds)
$1 = 200    (Step idle delay, milliseconds)
$2 = 255    (Step pulse invert, mask)
$3 = 255    (Step direction invert, mask)
$4 = 0    (Invert step enable pin, boolean)
$5 = 1    (Invert limit pins, boolean)
$6 = 1    (Invert probe pin, boolean)
$10 = 255    (Status report options, mask)
$11 = 2147483.648    (Junction deviation, millimeters)
$12 = 2147483.648    (Arc tolerance, millimeters)
$13 = 1    (Report in inches, boolean)
$20 = 0    (Soft limits enable, boolean)
$21 = 1    (Hard limits enable, boolean)
$22 = 1    (Homing cycle enable, boolean)
$23 = 255    (Homing direction invert, mask)
$24 = 2147483.648    (Homing locate feed rate, mm/min)
$25 = 2147483.648    (Homing search seek rate, mm/min)
$26 = 255    (Homing switch debounce delay, milliseconds)
$27 = 2147483.648    (Homing switch pull-off distance, millimeters)
$30 = 2147483648    (Maximum spindle speed, RPM)
$31 = 2147483648    (Minimum spindle speed, RPM)
$32 = 1    (Laser-mode enable, boolean)
$100 = 450.000    (X-axis travel resolution, step/mm)
$101 = 450.000    (Y-axis travel resolution, step/mm)
$102 = 450.000    (Z-axis travel resolution, step/mm)
$110 = 2147483.648    (X-axis maximum rate, mm/min)
$111 = 2147483.648    (Y-axis maximum rate, mm/min)
$112 = 2147483.648    (Z-axis maximum rate, mm/min)
$120 = 2147483.648    (X-axis acceleration, mm/sec^2)
$121 = 2147483.648    (Y-axis acceleration, mm/sec^2)
$122 = 2147483.648    (Z-axis acceleration, mm/sec^2)
$130 = 100.000    (X-axis maximum travel, millimeters)
$131 = 100.000    (Y-axis maximum travel, millimeters)
$132 = 100.000    (Z-axis maximum travel, millimeters)
ok

**Reply by rcecinati** [reply]
Joacim,
sorry for bothering, but I think I'm stuck with GRBL 0.9.  After searching around (it seems that github/grbl has definitely migrated to https://github.com/gnea/grbl/  and the ONLY version there is 1.1), I found at last this  one:
https://github.com/Denvi/Candle/releases
So I can download the Candle 0.9, it seems.
But what I find here is

Candle_0.9.tar.gz
21 MB 2016-10-08T19:31:44Z
Candle_0.9.zip
28.1 MB 2016-10-08T19:31:44Z
Source code (zip)
2016-10-08T19:26:47Z
Source code (tar.gz) 

Candle_0.9.tar.gz, Candle_0.9.zip  appear to be a pre-compiled version for Windows, and I need an Arduino platform. 
So i turn to source code 
Source code (zip) 
which does not seem to have the canonical .zip structure required by Arduino compilation (i've tried to tweak along, but no result). So, I try with manual compilation (as from https://docs.arduino.cc/software/ide-v1/tutorials/installing-libraries).  After decompressing and copying the src folder in my sketchbook libraries location, I can find in Sketch>Include library>Contributed Libraries > Candle-0.9. I need now to compile and upload, I think. But dependancies are requested:
C:\Users\rceci\Documents\Arduino\libraries\Candle-0.9/frmabout.h:7:10: fatal error: QDialog: No such file or directory
 #include <QDialog>

and I don't know where to pick them up. I am no expert in Arduino compiling, so can you kindly give me a clue, or point to and installable .zip library for Candle 0.9, much the same way I could get for 1.1?  Your help would be precious.
Thank you
Riccardo
          ^~~~~~~~~
compilation terminated.

**Reply by rcecinati** [reply]
latest news:
1) I understood that Candle is an application, not a version 0.9 of GRBL.
2) I've found in https://github.com/grbl/grbl-builds/blob/master/builds/grbl_v0_9j_atmega328p_16mhz_115200_build20160303.hex  the precompiled version of 0.9j
3) through XLoader (https://github.com/binaryupdates/xLoader), I've loaded to my Arduino
4) UGS now runs the Setup Wizard!  

Now I will run and test all the steps you suggested , first of all the wizard itself.
Thank you!

**Reply by breiler** [reply]
Yep, that is the one, good luck! =)

**Reply by breiler** [reply]
Oh, and the settings you posted shows that it's become corrupted, probably due to upgrading GRBL. I think that you should reset your configuration to the defaults using the command `$RST=*`

https://github.com/gnea/grbl/wiki/Grbl-v1.1-Commands#rst-rst-and-rst--restore-grbl-settings-and-data-to-defaults
